### ISSUE TITLE
Rules: sendCommand/postUpdate actions warning

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab-docs#185]
Add section to Rules to document warnings about using primitives with sendCommand/postUpdate actions

See https://community.openhab.org/t/sendcommand-documentation/9534/9

Question, that posting has a lot of background and generalizations to explain why using primitives with the command/update Actions is a problem. Should they be kept or should the docs update only include a succinct warning to watch out?

**Comment by LightIsLife** [comment]
As I am a newbie to github, I thought I'd share my thoughts on this issue first in a comment. 
I have developed some text that is based on Rich's post mentioned above.  As for the location, I would like to suggest to put this just before the Scripts section in the file on Rules (http://docs.openhab.org/configuration/rules-dsl.html#scripts). Anyway, here we go, let me know your thoughts. If it works, I can try pack this in a PR:

## Causing effects on items

Rules come to life by using input to cause something else to happen, to cause effects, such as change the state of an item and, e.g., switch on a light at sunset. The two most common commands that can change item values or states within rules are `sendCommand` and `postUpdate`. It is important to understand that they act slightly different and interact in specific ways with (event-based rule triggers)[{{base}}/configuration/rules-dsl.html#event-based-triggers]. For example `postUpdate` will will fire `received update`-based triggers, but not fire `received command`-based triggers. A `postUpdate` is often used to change the status of in item without causing any action to reflect any status changes that may be caused by other means (physically switch on a light). Correspondingly, `sendCommand` will fire `received command`-based triggers, but not fire `received update`-based triggers.

There are two possible syntaxes with which `sendCommand` and `postUpdate` can be used. There are seemingly small but important differences between the two. The two possible way to use `sendCommand` and `postUpdate` are:

Generic | Specific
-----------|-----------
`postUpdate(myItem, state_or_value)` | `myItem.postUpdate(state_or_value)`
`sendCommand(myItem, state_or_value)` | `myItem.sendCommand(state_or_value)`

And as for the answer which one to use, the easiest way is to remember that the syntax that will create the most stable code is the one in the column labeled “specific”: `myItem.postUpdate(state_or_value)` or `myItem.sendCommand(state_or_value)`. It provides by far the best option for avoiding most problems. This syntax ensures that any conversion (typing) of the `state_or_value` is done in a way that is most suitable for `myItem`. 

### For those who want to have more background, here the long answer:
The reasons as to why `myItem.sendCommand` and `myItem.postUpdate` are the preferred syntax lie within Java, the object-oriented program languages on which openHab and its items, rules, etc are built. Java and the Rules DSL have two basic types, primitives and objects. Whenever there is a type after a `var` or a `val` statement with lower-case first letter (e.g., `var int` or `var boolean`) we are dealing with a primitive type. Whenever the `val` and `var` are followed with type in upper-case (e.g., `var String`, `var Number`), the type is an object. This means that items are objects. Objects are more complex than simple variables and have methods associated with them that enable interaction with objects, as for example changing the state of an item. The syntax `myitem.sendCommand` or `myitem.postUpdate` ensures that the most appropriate and best method to convert/type the `state_or_value` into the structure that is needed by the object (here `myItem`). 

The alternative syntax, `sendCommand(myItem, state_or_value)` applies only a set of generic methods, which are not able to accommodate all cases of conversion that may be necessary to make sure that `state_or_value` can be applied to `myItem`. 

For example, if `state_or_value` is typed as a primitive (e.g., `var int my_number = 3`) and myItem is of the type Dimmer (note the capitalized ‘D’ denoting an object: 
* the following command ***will fail***: ~~sendCommand(myItem, my_number)~~. 
* However, the following command **will work**: `myItem.sendCommand(my_number)`. 

### Even more details:

As all object-oriented computer languages, Java and the Rules DSL have implemented the concept of Inheritance. However, inheritance only applies to Objects and does **not** apply to primiteves. Inheritance allows to take an existing Object type, called a Class, and adding to it to make it into something different, but it still can do everything the existing Object could do. This “something different” becomes a Child of the original Class. The top level base Class for all Objects in Java and the Rules DSL is called Object. The Child can be treated as if it were the parent because everything the parent can do the child can do too.

In addition to other useful things, Object implements a method called `toString`. And since Object is the parent of all Objects, that means all Classes also implement a `toString` method. _But primitives do not inherit from Object. They don't inherit from anything and they don't have any methods at all which includes toString._

The `sendCommand` is a generic action and therefore, by definition, needs to be able to work with all Objects. A closer look reveals that this action only supports two arguments: String, String, because all Objects will support the conversion `toString`. But a Primitive is not an object and has no `toString` (in fact a Primitive has no methods at all), as a consequence in almost all cases it cannot convert the primitive value to a String and the use of `sendCommand(myItem, primitive)`, with the use of a primitive as the second argument, will fail. 

With regard to the `sendCommand` method on the Items, here the `sendCommand` method is defined by the class for that Item which allows the class to create `sendCommand` methods that take different types of data. For example, the `NumberItem` class would have a `sendCommand(int)`, `sendCommand(long)`, `sendCommand(float)`, `sendCommand(double)`, `sendCommand(Number)`, `sendCommand(DecimalType)`, and `sendCommand(String)` methods. Each of these separate methods is individually written to handle all of these different types of Objects and primitives unlike the Action which can only handle Strings.

In a nutshell, using the syntax `myItem.sendCommand(state_or_value)` or `myItem.sendUpdate(state_or_value)` will help aavoid many problems that the generic method may cause.

**Comment by ThomDietrich** [comment]
Hey @LightIsLife,
thanks for that comprehensive contribution on this **important** topic. I'd say you can open the PR right away. As you said you are new to github, Here's the short explanation on how to do edits directly from the GitHub page:

1. Go to the rules sourcecode page: https://github.com/openhab/openhab-docs/blob/gh-pages/configuration/rules-dsl.md
2. Click the small "Edit this file" pen on the top right
3. Incorporate your changes
4. Give your addition a meaningful name and 
  * Important: [Sign of your work](https://github.com/openhab/openhab-docs/blob/gh-pages/CONTRIBUTING.md#sign-your-work)
5. Follow the process to open a PR.

That should be it. Let us know if there are issues, also don't worry about mistakes, everything can still be improved after the PR was created.

Here are a few first comments from my side, which we can best further discuss after you've created the PR:

* The first paragraph includes a few important details but I think we should make the difference crystal clear before going into details. I think most users will only be interested in the first 1-2 sentences giving them an idea of wether to use sendCommand or postUpdate.

* Please do not mix explanation of events and event-based triggers. The fact, that "`postUpdate` will not fire `received command` triggers" should be described in the respective section of the rules article. What do you think?

* In the short version of the article I'd clearly recommend the `myItem.postUpdate(state_or_value)` method instead of showing them side-by-side.

* The more detailed background info on the topic might better placed at the end of the article to not clutter up the article.

**Comment by LightIsLife** [comment]
Thanks much @ThomDietrich, I appreciate the encouragement and help. If I did everything correctly, PR#469 should be the one. 
Thanks also for your comments, I worked a little more on the text, hope we are getting closer. Happy to continue working on it.
